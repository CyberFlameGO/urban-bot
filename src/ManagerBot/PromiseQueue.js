export class PromiseQueue {
    constructor() {
        this.last = Promise.resolve();
    }

    next(callback) {
        this.last = this.last.then(callback);
        return this.last;
    }
}
