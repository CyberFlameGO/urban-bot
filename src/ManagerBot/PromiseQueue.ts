import { AnyFunction } from '../types/common';

export class PromiseQueue {
    last: Promise<any>;

    constructor() {
        this.last = Promise.resolve();
    }

    next(callback: AnyFunction) {
        this.last = this.last.then(callback);
        return this.last;
    }
}
