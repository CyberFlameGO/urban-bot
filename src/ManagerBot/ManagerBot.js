import { PromiseQueue } from './PromiseQueue';

export class ManagerBot {
    constructor(bot) {
        this.bot = bot;
        this.promiseQueueMap = new Map();
    }

    addChat(id) {
        this.promiseQueueMap.set(id, new PromiseQueue());
    }

    deleteChat(id) {
        this.promiseQueueMap.delete(id);
    }

    on(event, listener, chatId) {
        return this.bot.on(event, function(ctx) {
            if (chatId !== undefined) {
                const { id: chatIdFromEvent } = ctx.chat;

                if (chatId === chatIdFromEvent) {
                    listener(ctx);
                }

                return;
            }

            listener(ctx);
        });
    }

    emit(type, message) {
        return this.bot.emit(type, message);
    }

    removeListener(eventName, listener) {
        return this.bot.removeListener(eventName, listener);
    }

    sendMessage(nodeName, chatId, data) {
        const promiseQueueByChatId = this.promiseQueueMap.get(chatId);

        if (promiseQueueByChatId === undefined) {
            throw new Error('Specify chatId before send message via managerBot.addChat(chatId)');
        }

        return promiseQueueByChatId.next(() => {
            return this.bot.sendMessage(nodeName, chatId, data);
        });
    }

    updateMessage(nodeName, chatId, data, meta) {
        return this.bot.updateMessage(nodeName, chatId, data, meta);
    }

    deleteMessage(nodeName, chatId, data, meta) {
        return this.bot.deleteMessage(nodeName, chatId, data, meta);
    }
}
