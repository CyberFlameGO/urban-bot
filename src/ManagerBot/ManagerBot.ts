import { PromiseQueue } from './PromiseQueue';
import EventEmitter from 'events';
import { ProcessUpdate, UrbanBot } from '../types/UrbanBot';
import { UrbanExistingMessage, UrbanMessage } from '../types/Messages';
import { UrbanListener } from '../types';
import { UrbanEvent } from '../types/Events';

type Chat = {
    eventEmitter: EventEmitter;
    promiseQueue: PromiseQueue;
};

export class ManagerBot<Type = unknown, NativeEventPayload = unknown, Meta = unknown> {
    private chats = new Map<string, Chat>();
    private eventEmitter: EventEmitter;

    constructor(private bot: UrbanBot<Type, NativeEventPayload, Meta>) {
        this.eventEmitter = new EventEmitter();

        bot.initializeProcessUpdate(this.processUpdate);
    }

    processUpdate: ProcessUpdate<Type, NativeEventPayload> = (event) => {
        this.eventEmitter.emit('any', event);
        this.eventEmitter.emit(event.type);

        const { id } = event.chat;
        const chat = this.chats.get(id);

        if (chat === undefined) {
            throw new Error('Specify chatId via managerBot.addChat(chatId) to process messages for specific chat');
        }

        chat.eventEmitter.emit('any', event);
        chat.eventEmitter.emit(event.type, event);
    };

    addChat(id: string) {
        this.chats.set(id, { promiseQueue: new PromiseQueue(), eventEmitter: new EventEmitter() });
    }

    deleteChat(id: string) {
        this.chats.delete(id);
    }

    on<Event extends UrbanEvent<Type, NativeEventPayload> = UrbanEvent<Type, NativeEventPayload>>(
        eventName: Event['type'] | 'any',
        listener: UrbanListener<Event>,
        chatId?: string,
    ) {
        if (chatId === undefined) {
            return this.eventEmitter.on(eventName, listener);
        } else {
            const chat = this.chats.get(chatId);
            if (chat === undefined) {
                throw new Error(
                    'Specify chatId via managerBot.addChat(chatId) to subscribe on messages for specific chat',
                );
            }

            return chat.eventEmitter.on(eventName, listener);
        }
    }

    emit<Event extends UrbanEvent<Type, NativeEventPayload> = UrbanEvent<Type, NativeEventPayload>>(
        eventName: Event['type'] | 'any',
        event: Event,
    ) {
        this.eventEmitter.emit('any', event);
        this.eventEmitter.emit(eventName, event);
    }

    removeListener<Event extends UrbanEvent<Type, NativeEventPayload> = UrbanEvent<Type, NativeEventPayload>>(
        eventName: Event['type'] | 'any',
        listener: UrbanListener<Event>,
        chatId?: string,
    ) {
        if (chatId === undefined) {
            return this.eventEmitter.removeListener(eventName, listener);
        } else {
            const chat = this.chats.get(chatId);
            if (chat === undefined) {
                throw new Error('Specify chatId via managerBot.addChat(chatId) to sendMessage for specific chat');
            }

            return chat.eventEmitter.removeListener(eventName, listener);
        }
    }

    sendMessage(message: UrbanMessage): Promise<Meta> {
        const chatById = this.chats.get(message.chat.id);

        if (chatById === undefined) {
            throw new Error('Specify chatId via managerBot.addChat(chatId) to sendMessage for specific chat');
        }

        return chatById.promiseQueue.next(() => {
            return this.bot.sendMessage(message);
        });
    }

    updateMessage(message: UrbanExistingMessage<Meta>) {
        return this.bot.updateMessage(message);
    }

    deleteMessage(message: UrbanExistingMessage<Meta>) {
        return this.bot.deleteMessage(message);
    }
}
