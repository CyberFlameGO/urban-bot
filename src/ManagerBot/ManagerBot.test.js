import { ManagerBot } from './ManagerBot';

const testBot = {
    on: jest.fn(),
    emit: jest.fn(),
    sendMessage: jest.fn(),
    removeListener: jest.fn(),
    updateMessage: jest.fn(),
    deleteMessage: jest.fn(),
};

const testBotEmitter = {
    on(event, listener) {
        this[event] = listener;
    },
    emit(event, data) {
        this[event](data);
    },
};

describe('ManagerBot', () => {
    describe('method on', () => {
        const event = 'test-event';
        const listener = jest.fn();

        afterEach(() => {
            listener.mockClear();
        });

        it('call inner bot method on', () => {
            const managerBot = new ManagerBot(testBot);

            managerBot.on(event, listener);

            expect(testBot.on).toHaveBeenCalledTimes(1);
            expect(testBot.on.mock.calls[0][0]).toBe(event);
            expect(testBot.on.mock.calls[0][1]).not.toBe(listener);
        });

        it('call listener if chatId is empty', () => {
            const managerBot = new ManagerBot(testBotEmitter);

            managerBot.on(event, listener);
            const data = { chat: { id: 123 } };
            const data2 = { chat: { id: 345 } };

            testBotEmitter.emit(event, data);

            expect(listener).toHaveBeenCalledTimes(1);
            expect(listener).toHaveBeenLastCalledWith(data);

            testBotEmitter.emit(event, data2);

            expect(listener).toHaveBeenCalledTimes(2);
            expect(listener).toHaveBeenLastCalledWith(data2);
        });

        it('call listener if chatId is the same with emit data', () => {
            const managerBot = new ManagerBot(testBotEmitter);

            const id = 123;

            managerBot.on(event, listener, id);
            const data = { chat: { id } };

            testBotEmitter.emit(event, data);

            expect(listener).toHaveBeenCalledTimes(1);
            expect(listener).toHaveBeenLastCalledWith(data);
        });

        it("doesn't call listener if chatId is not the same with emit data", () => {
            const managerBot = new ManagerBot(testBotEmitter);

            const eventId = '123';
            const id = 123;
            const id2 = 345;

            managerBot.on(event, listener, eventId, id);
            const data = { chat: { id: id2 } };

            testBotEmitter.emit(event, data);

            expect(listener).toHaveBeenCalledTimes(0);
        });
    });

    it('method emit', () => {
        const managerBot = new ManagerBot(testBot);

        const event = 'test-event';
        const listener = () => {};

        managerBot.emit(event, listener);

        expect(testBot.emit).toHaveBeenCalledTimes(1);
        expect(testBot.emit).toHaveBeenLastCalledWith(event, listener);

        managerBot.emit(event, listener);

        expect(testBot.emit).toHaveBeenCalledTimes(2);
    });

    it('method removeListener', () => {
        const managerBot = new ManagerBot(testBot);

        const event = 'test-event';
        const listener = () => {};
        const eventId = '123';

        managerBot.removeListener(event, listener, eventId);

        expect(testBot.removeListener).toHaveBeenCalledTimes(1);
        expect(testBot.removeListener).toHaveBeenLastCalledWith(event, listener, eventId);

        managerBot.removeListener(event, listener, eventId);

        expect(testBot.removeListener).toHaveBeenCalledTimes(2);
    });

    it('method updateMessage', () => {
        const managerBot = new ManagerBot(testBot);

        const nodeName = 'text';
        const chatId = 123;
        const data = { text: '123' };
        const meta = { messageId: 345 };
        managerBot.updateMessage(nodeName, chatId, data, meta);

        expect(testBot.updateMessage).toHaveBeenCalledTimes(1);
        expect(testBot.updateMessage).toHaveBeenLastCalledWith(nodeName, chatId, data, meta);

        managerBot.updateMessage(nodeName, chatId, data, meta);

        expect(testBot.updateMessage).toHaveBeenCalledTimes(2);
    });

    it('method deleteMessage', () => {
        const managerBot = new ManagerBot(testBot);

        const nodeName = 'text';
        const chatId = 123;
        const data = { text: '123' };
        const meta = { messageId: 345 };
        managerBot.deleteMessage(nodeName, chatId, data, meta);

        expect(testBot.deleteMessage).toHaveBeenCalledTimes(1);
        expect(testBot.deleteMessage).toHaveBeenLastCalledWith(nodeName, chatId, data, meta);

        managerBot.deleteMessage(nodeName, chatId, data, meta);

        expect(testBot.deleteMessage).toHaveBeenCalledTimes(2);
    });

    describe('method sendMessage', () => {
        afterEach(() => {
            testBot.sendMessage.mockClear();
        });

        const nodeName = 'text';
        const chatId = 123;
        const data = { text: '123' };

        const nodeName2 = 'img';
        const data2 = { src: 'http://something' };

        it('send one message', () => {
            const managerBot = new ManagerBot(testBot);

            managerBot.addChat(chatId);

            testBot.sendMessage.mockReturnValue(Promise.resolve());

            const res = managerBot.sendMessage(nodeName, chatId, data);

            return res.then(() => {
                expect(testBot.sendMessage).toHaveBeenCalledTimes(1);
                expect(testBot.sendMessage).toHaveBeenLastCalledWith(nodeName, chatId, data);
            });
        });

        it('send two messages', () => {
            const managerBot = new ManagerBot(testBot);

            managerBot.addChat(chatId);

            testBot.sendMessage.mockReturnValue(Promise.resolve());

            const res1 = managerBot.sendMessage(nodeName, chatId, data);
            const res2 = managerBot.sendMessage(nodeName2, chatId, data2);

            return Promise.all([res1, res2]).then(() => {
                expect(testBot.sendMessage).toHaveBeenCalledTimes(2);
                expect(testBot.sendMessage).toHaveBeenCalledWith(nodeName, chatId, data);
                expect(testBot.sendMessage).toHaveBeenCalledWith(nodeName2, chatId, data2);
            });
        });

        it('send next message only when previous was finished', () => {
            const managerBot = new ManagerBot(testBot);

            managerBot.addChat(chatId);

            const firstPromise = managerBot.chats.get(chatId).last;

            testBot.sendMessage.mockReturnValueOnce(
                new Promise((resolve) => {
                    setTimeout(resolve, 1000);
                }),
            );
            const res1 = managerBot.sendMessage(nodeName, chatId, data);

            testBot.sendMessage.mockReturnValueOnce(Promise.resolve());
            managerBot.sendMessage(nodeName2, chatId, data2);

            return firstPromise
                .then(() => {
                    expect(testBot.sendMessage).toHaveBeenNthCalledWith(1, nodeName, chatId, data);
                    expect(testBot.sendMessage).not.toHaveBeenCalledTimes(2);

                    return res1;
                })
                .then(() => {
                    expect(testBot.sendMessage).toHaveBeenNthCalledWith(2, nodeName2, chatId, data2);
                });
        });

        it('throw error without initializing chatId', () => {
            const managerBot = new ManagerBot(testBot);

            expect(() => managerBot.sendMessage(nodeName, chatId, data)).toThrowErrorMatchingSnapshot();
        });

        it('throw error if send message after delete chatId', async () => {
            const managerBot = new ManagerBot(testBot);
            managerBot.addChat(chatId);
            testBot.sendMessage.mockReturnValue(Promise.resolve());

            await managerBot.sendMessage(nodeName, chatId, data);
            expect(testBot.sendMessage).toHaveBeenCalledTimes(1);

            managerBot.deleteChat(chatId);
            expect(() => managerBot.sendMessage(nodeName2, chatId, data2)).toThrowErrorMatchingSnapshot();
        });
    });
});
