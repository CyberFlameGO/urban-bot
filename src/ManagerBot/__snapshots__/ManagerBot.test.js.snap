// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManagerBot method sendMessage throw error without initializing chatId 1`] = `"Specify chatId before send message via managerBot.addChat(chatId)"`;
